### PR TITLE
Improve documentation of EntityId

### PR DESCRIPTION
**Pull request by brightbyte**
This patch documents the intended semantics of the methods in the
EntityId base class, to align them to the behavior defined in
docs/foreign-entity-ids.wiki

### CODE REVIEW

**Comment by brightbyte** [comment]
Ping @thiemowmde @manicki @filbertkm

File: docs/foreign-entity-ids.wiki [hunk 1]
```diff
@@ -17,7 +17,7 @@ When receiving prefixed entities from another repository, prefixes are "chained"
 
 * When reading <code>d:Q5</code> from repository <code>foo</code>, it is turned into <code>foo:d:Q5</code>, meaning ''<code>Q5</code> at the repo that repo <code>foo</code> calls <code>d</code>''
 * The local repository may have mappings defined for the prefixes used by other wikis, e.g. <code>foo:d</code> would be known to be the same as the local prefix <code>wd</code> for Wikidata. Mappings are defined as a two-dimensional array, e.g. part of mapping definition for example mentioned here would look like below:
-<code> [ ... 'foo' => [ 'd' => 'wd' ], ... ]</code>
+<code> array( ... 'foo' => array( 'd' => 'wd' ), ... )</code>
```
Comment (manicki) [inline docs/foreign-entity-ids.wiki]: Is this change intended?

File: src/Entity/EntityId.php [hunk 2]
```diff
@@ -43,12 +85,16 @@ private static function assertValidSerialization( $serialization ) {
 	}
 
 	/**
-	 * @return string
+	 * @return string The type of the designated Entity.
```
Comment (manicki) [inline src/Entity/EntityId.php]: This description could be omitted I believe.

File: src/Entity/EntityId.php [hunk 3]
```diff
@@ -7,6 +7,38 @@
...
+ * the EntityId object was instantiated in, or, for Wikibase  clients,
+ * the default repo of the client the EntityId object was instantiated in.
+ *
+ * A repository can be specified in the string representation of an EntityId
+ * as a prefix, separated by a colon (":") from the local part of the ID:
+ * "<repo>:<local-id>". Multiple such prefixes can be "chained", as in
+ * "foo:bar:rudd:X123"; in that case, only the first prefix ("foo") is
+ *  considered the repository name, the remaining prefixes are part of the
+ * "local part" of the ID ("bar:rudd:X123"), since they are to be resolved
+ * "locally" in the context of the context given as the first prefix.
```
Comment (manicki) [inline src/Entity/EntityId.php]: "in the context of the **repository** given..."

File: src/Entity/EntityId.php [hunk 4]
```diff
@@ -104,8 +150,18 @@ public static function joinSerialization( array $parts ) {
 	}
 
 	/**
-	 * Returns '' for local IDs and the foreign repository name for foreign IDs. For chained IDs (e.g. foo:bar:Q42) it
-	 * will return only the first part.
+	 * Returns the name of the repository which serves as the context for resolving this EntityId.
+	 * This method is the complement of getLocalPart().
+	 *
+	 * For local IDs with no repository name attached, getRespoitoryName() will return the empty
```
Comment (manicki) [inline src/Entity/EntityId.php]: getRepositoryName mispelled slightly

File: src/Entity/EntityId.php [hunk 5]
```diff
@@ -7,6 +7,38 @@
...
+ * which they can be resolved. IDs that do not provide a repository name
+ * (or, technically, have the repository name set to the empty string, '')
+ * are referred to as "local". IDs that do provide a repository name are
+ * referred to as "foreign" (even if they, by some detour, are equivalent to
+ * a local ID).
+ *
+ * The intended interpretation of "local" IDs is that they can be resolved
+ * in the default context - that is, they belong to the Wikibase repository
+ * the EntityId object was instantiated in, or, for Wikibase  clients,
+ * the default repo of the client the EntityId object was instantiated in.
```
Comment (thiemowmde) [inline src/Entity/EntityId.php]: This is confusing and I believe even wrong. What this sentence basically says is that the interpretation of "local" is meaningless. Could mean anything you want. May not even be "local". Also it's unclear to what "instantiated" refers to if an ID is, for example, read from a database or dump. Same for "default". To what does this refer to? Who defines said "default"? Who knows about it? Certainly not the DataModel.

**Comment by brightbyte** [comment]
> Furthermore, this component does not need this knowledge.

The data model needs to provide a specification for the interpretation of IDs. That specification needs to contain all information required by application code (including 3rd party clients) to successfully resolve an EntityId. 

Even if the code in this component does not need this knowledge, this component needs to define the respective contract, so that different users of the data model are interoperable.

**Comment by brightbyte** [comment]
For the record, I don't insist on the terms "local" and "foreign". Maybe "relative" and "qualified" are better. I do maintain that it's important to clearly define the meaning of prefixes in general, and the empty prefix in particular, here in this component.

I agree that this component should not know about wikis, or their "location". But it needs to know about "repositories" as an abstract concept: a repo is a storage location, name resolution context, and identifier namespace. We need to make use of this concept in the specification of the data model. It's the only way to make sure all users of this component share the same interpretation of the data.

**Comment by thiemowmde** [comment]
"Local" and "foreign" **are** locations.

In parallel I'm in an email thread with @brightbyte and it appears we have two different ideas of what this component represents. I believe this is the PHP implementation of Wikibase' base DataModel and not a conceptual data model that describes the interpretation of details like what an empty repository prefix in the serializations means.

**Comment by JeroenDeDauw** [comment]
General observation here, not entirely specific to this PR: I'm not happy with foundational concepts being made more complicated. Needed to have long documentation on value objects is also kinda disconcerting.

**Comment by JeroenDeDauw** [comment]
Please reopen if this should still show up in the open PR list
